Initial bindings - Table and Route services

mjkillough opened the pull request:
This adds initial bindings for the Table and Route services.

The OSRM API uses a lot of C++ templates, so it is not easy to generate Rust bindings for it. The templates confuse [`bindgen`](https://github.com/rust-lang/rust-bindgen), so we'd be left creating them by hand. This instead uses the [`libosrmc`](https://github.com/daniel-j-h/libosrmc) library, which is a simple C wrapper around the C++ API, and then builds a Rust API on top of that.

The `osrmc-sys` crate should be complete, but there are other services we can add safe wrappers to in the `osrm-rs` bindings if we'd like.

Some things I haven't done:
 - I haven't added any tests, as it isn't clear how we can best test this. If we download a random OSM map, how do we assert the routes are correct?
 - No documentation yet, but we can come back and add that later.

mjkillough committed:
Add empty crates.

mjkillough committed:
Add basic OSRM Table API.

Commit:
Add README.md

Commit:
Expose TableResponse.

Commit:
Use HTTPS URL for Git Submodule.

We don't need write access on it, and HTTPS is more likely to work for
those who don't have SSH configured.

Commit:
Add support for Route service.

Commit:
Change boost libs - -mt no longer required.

It used to mean 'multi-threaded', but apparently that's enabled by
default these days. Ubuntu doesn't have the `-mt` version of the libs
and we see linker errors.

Commit:
Remove stutter in type names.

Commit:
Update README.

Commit:
Add LICENSE.

Commit:
s/Response/RouteResponse/g

Matches TableResponse.

Commit:
s/OSRM/Osrm/g

Rust prefers acronyms to still follow UpperCamelCase.

Commit:
Boost libraries differ on macOS/Ubuntu.

Commit:
Update README.

Commit:
Update .gitignore.

Commit:
s/Point/Coordinate/g

Commit:
Make OsrmcError private.

Commit:
Impl Send and Sync.

me commented on src/lib.rs:
```diff
@@ -0,0 +1,108 @@
+
+    #[test]
+    fn it_works() {
+        let osrm = Osrm::new("./data/1.osrm").expect("uh oh");
```
`osrm-c` has a Makefile to download an `.osm` file and make it into `.osrm` for testing. Should we have something similar?

mjkillough replied:
Yeah, we can do something similar. I might check the returned duration is non-zero, but not much more. 

(I wasn't aware of osrm-c. I think libosrmc might be a little bit more 'official', but it's good to know there are alternatives!)

Antti commented on README.md:
````diff
@@ -0,0 +1,34 @@
+
+```sh
+# Install OSRM and its dependencies.
+brew install boost tbb osrm-backend
````
Wouldn't brew install dependencies automatically?

mjkillough replied:
Good point!

Commit:
Add tests.

Provide script `./prepare-test-data.sh` which downloads and processes
Berlin, and some tests that check we produce non-zero data.

Commit:
Add .travis.yml.

Commit:
Update README.

Commit:
Tweak .travis.yml.

me commented on README.md:
````diff
@@ -0,0 +1,47 @@
+
+```
+# First download/process the required maps:
+./prepare-test-data
````
Missing `.sh`?

me reviewed (approved):
LGTM!

Commit:
Make Error Sync.

It's a `std::string`, and we only ever read it.

Commit:
Make paths easier to deal with.
